paasta_tools/api/views/instance.py

```diff
@@ -336,37 +356,86 @@ def marathon_smartstack_status(
+            location_dict = smartstack_tools.build_smartstack_location_dict(
+                location,
+                matched_smartstack_backends_and_tasks,
+                should_return_individual_backends,
+            )
+            service_mesh_status["locations"].append(location_dict)
+        elif service_mesh == ServiceMesh.ENVOY:
+            envoy_host = hosts[0]
+            backends = envoy_tools.get_backends(
                 registration,
-                synapse_host=synapse_host,
-                synapse_port=settings.system_paasta_config.get_synapse_port(),
-                synapse_haproxy_url_format=settings.system_paasta_config.get_synapse_haproxy_url_format(),
-            ),
-            key=lambda backend: backend["status"],
-            reverse=True,  # put 'UP' backends above 'MAINT' backends
-        )
-        matched_backends_and_tasks = match_backends_and_tasks(sorted_backends, tasks)
-        location_dict = smartstack_tools.build_smartstack_location_dict(
-            location, matched_backends_and_tasks, should_return_individual_backends
-        )
-        smartstack_status["locations"].append(location_dict)
+                envoy_host=envoy_host,
+                envoy_admin_port=settings.system_paasta_config.get_envoy_admin_port(),
+            )
+            sorted_envoy_backends = sorted(
+                [backend[0] for backend in backends],
+                key=lambda backend: backend["eds_health_status"],
+            )
+            casper_proxied_backends = {
+                (backend["address"], backend["port_value"])
+                for backend, is_casper_proxied_backend in backends
+                if is_casper_proxied_backend
+            }
+
+            matched_envoy_backends_and_tasks = envoy_tools.match_backends_and_tasks(
+                sorted_envoy_backends, tasks,
+            )
 
-    return smartstack_status
+            running_backends_count = 0
+            envoy_backends = []
+            is_proxied_through_casper = False
+            for backend, task in matched_envoy_backends_and_tasks:
+                if backend is None:
+                    continue
+                if backend["eds_health_status"] == "HEALTHY":
+                    running_backends_count += 1
+                if should_return_individual_backends:
+                    backend["has_associated_task"] = task is not None
+                    envoy_backends.append(backend)
+                if (
+                    backend["address"],
+                    backend["port_value"],
+                ) in casper_proxied_backends:
+                    is_proxied_through_casper = True
+            location_dict = {
+                "name": location,
+                "running_backends_count": running_backends_count,
+                "backends": envoy_backends,
+                "is_proxied_through_casper": is_proxied_through_casper,
+            }
+            service_mesh_status["locations"].append(location_dict)
```

Comment:
can we make this a separate function, too?  I am concerned that this function is getting pretty complex.